src/types/types.zig

```diff
@@ -231,17 +231,78 @@ pub const Num = union(enum) {
         pub fn placeholder() Compact {
             return Compact{ .int = .u8 };
         }
+    };
+
+    /// the Frac data type
+    pub const Frac = struct {
+        /// the precision of a frac
+        pub const Precision = enum {
+            f32,
+            f64,
+            dec,
+
+            /// Get the lowest precision needed to hold the provided float
+            /// This only supports f32s and f64s. Decimals must be assigned based on different criteria.
+            /// When in doubt, this prefers f32s.
+            pub fn fromValue(value: f64) Frac.Precision {
+                if (std.math.isNan(value) or std.math.isInf(value)) {
+                    return .f32;
+                }
 
-        /// the Frac data type
-        pub const Frac = struct {
-            /// the precision of a frac
-            pub const Precision = enum { f32, f64, dec };
+                // Check if the value fits in f32 range and precision
+                if (@abs(value) == 0.0) {
+                    return .f32;
+                } else if (value <= std.math.floatMax(f32) and value >= std.math.floatMin(f32)) {
+                    // Not every f64 can be downcast to f32 with the same precision
+                    const as_f32 = @as(f32, @floatCast(value));
+                    const back_to_f64 = @as(f64, @floatCast(as_f32));
+                    if (value == back_to_f64) {
```

Comment:
should this be not equals?

Reply:
It should be equals! I update the code comment to make this a little more clear!